File: sorting in Arrays/JAVA/MergeSort.java
import java.util.Arrays;

public class MergeSort {
    private void merge(int[] arr, int left, int mid, int right){

        int[] L = Arrays.copyOfRange(arr, left, mid+1);
        int[] R = Arrays.copyOfRange(arr, mid+1, right+1);
        int i = 0;
        int j = 0;
        int k = left;

        while(i<L.length && j<R.length){
            if(L[i]< R[j]){
                arr[k] = L[i];
                k++;
                i++;
            } 
            else{
                arr[k] = R[j];
                k++;
                j++;
            }
        while(i< L.length){
            arr[k] = L[i];
            i++;
            k++;
        }
        while(j< R.length){
            arr[k] = R[j];
            j++;
            k++;
        }



        }
    }

    public void sortArray(int[] arr, int left, int right){
        if (left >= right){
            return;
        }
        int mid = (int)((left+right)/2);
        sortArray(arr, left, mid);
        sortArray(arr, mid+1, right);
        merge(arr, left, mid, right);

    }
    public static void main(String[] args) {
        MergeSort mergeSort = new MergeSort();
        int[] arr = {2, 4, 1, 3, 5};
        mergeSort.sortArray(arr, 0,  arr.length - 1);

        for(int i=0; i<arr.length; i++){
            System.out.println(arr[i]);
        }
        
    }
    
}

File: sorting in Arrays/Python/merge_sort.py
class MergeSort:

    def merge(self, arr, left, mid, right):
        # create temp arrays
        L = arr[left:mid+1]
        R = arr[mid+1:right+1]

        i = 0
        j = 0
        k = left

        # merge
        while i < len(L) and j < len(R):
            if L[i] <= R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1

        # remaining elements
        while i < len(L):
            arr[k] = L[i]
            i += 1
            k += 1

        while j < len(R):
            arr[k] = R[j]
            j += 1
            k += 1

    def sortArray(self, arr, left, right):
        if left >= right:      # base condition
            return

        mid = (left + right) // 2

        self.sortArray(arr, left, mid)
        self.sortArray(arr, mid + 1, right)
        self.merge(arr, left, mid, right)

if __name__ == "__main__":
    arr = [2, 4, 1, 3, 5]
    MergeSort().sortArray(arr, 0, len(arr)-1)
    print(arr)